kernel/typeops.ml

```diff
@@ -649,113 +649,415 @@ let check_let_annot env s x c t =
+
+  module Tbl : Hashtbl.S with type key = t
+
+end = struct
+
+(* copied from Constr.hasheq *)
+let hasheq t1 t2 =
+  let array_eqeq a b = Array.equal (==) a b in
+  let invert_eqeq iv1 iv2 =
+    match iv1, iv2 with
+    | NoInvert, NoInvert -> true
+    | NoInvert, CaseInvert _ | CaseInvert _, NoInvert -> false
+    | CaseInvert {indices=i1}, CaseInvert {indices=i2} ->
+      i1 == i2
+  in
+  let hasheq_ctx (nas1, c1) (nas2, c2) =
+    array_eqeq nas1 nas2 && c1 == c2
+  in
+
+  match kind t1, kind t2 with
+    | Rel n1, Rel n2 -> n1 == n2
+    | Meta m1, Meta m2 -> m1 == m2
+    | Var id1, Var id2 -> id1 == id2
+    | Sort s1, Sort s2 -> s1 == s2
+    | Cast (c1,k1,t1), Cast (c2,k2,t2) -> c1 == c2 && k1 == k2 && t1 == t2
+    | Prod (n1,t1,c1), Prod (n2,t2,c2) -> n1 == n2 && t1 == t2 && c1 == c2
+    | Lambda (n1,t1,c1), Lambda (n2,t2,c2) -> n1 == n2 && t1 == t2 && c1 == c2
+    | LetIn (n1,b1,t1,c1), LetIn (n2,b2,t2,c2) ->
+      n1 == n2 && b1 == b2 && t1 == t2 && c1 == c2
+    | App (c1,l1), App (c2,l2) -> c1 == c2 && array_eqeq l1 l2
+    | Proj (p1,r1,c1), Proj(p2,r2,c2) -> p1 == p2 && r1 == r2 && c1 == c2
+    | Evar (e1,l1), Evar (e2,l2) -> e1 == e2 && SList.equal (==) l1 l2
+    | Const (c1,u1), Const (c2,u2) -> c1 == c2 && u1 == u2
+    | Ind (ind1,u1), Ind (ind2,u2) -> ind1 == ind2 && u1 == u2
+    | Construct (cstr1,u1), Construct (cstr2,u2) -> cstr1 == cstr2 && u1 == u2
+    | Case (ci1,u1,pms1,(p1,r1),iv1,c1,bl1), Case (ci2,u2,pms2,(p2,r2),iv2,c2,bl2) ->
+      (** FIXME: use deeper equality for contexts *)
+      u1 == u2 && array_eqeq pms1 pms2 &&
+      ci1 == ci2 && hasheq_ctx p1 p2 && r1 == r2 &&
+      invert_eqeq iv1 iv2 && c1 == c2 && Array.equal hasheq_ctx bl1 bl2
+    | Fix ((ln1, i1),(lna1,tl1,bl1)), Fix ((ln2, i2),(lna2,tl2,bl2)) ->
+      Int.equal i1 i2
+      && Array.equal Int.equal ln1 ln2
+      && array_eqeq lna1 lna2
+      && array_eqeq tl1 tl2
+      && array_eqeq bl1 bl2
+    | CoFix(ln1,(lna1,tl1,bl1)), CoFix(ln2,(lna2,tl2,bl2)) ->
+      Int.equal ln1 ln2
+      && array_eqeq lna1 lna2
+      && array_eqeq tl1 tl2
+      && array_eqeq bl1 bl2
+    | Int i1, Int i2 -> i1 == i2
+    | Float f1, Float f2 -> Float64.equal f1 f2
+    | Array(u1,t1,def1,ty1), Array(u2,t2,def2,ty2) ->
+      u1 == u2 && def1 == def2 && ty1 == ty2 && array_eqeq t1 t2
+    | (Rel _ | Meta _ | Var _ | Sort _ | Cast _ | Prod _ | Lambda _ | LetIn _
+      | App _ | Proj _ | Evar _ | Const _ | Ind _ | Construct _ | Case _
+      | Fix _ | CoFix _ | Int _ | Float _ | Array _), _ -> false
+
+(* local_env: can't use actual local env unless we start taking
+   Environ.env to deal with Case
+   instead use physical equality and make fresh int refs
+*)
+type local_env = int ref
+
+module Self = struct
+
+type t = {
+  local_env : local_env;
+  self : constr;
+  kind : (t,t,Sorts.t,UVars.Instance.t,Sorts.relevance) kind_of_term;
+  hash : int;
+  mutable refcount : int;
+}
+
+let equal a b =
+  a.local_env == b.local_env
```

Comment:
this is probably too coarse, doing some analysis of unused variables could help
eg if we have `((fun _ (* unused *) => big), (fun _ (* unused *) => big))` the current code still double checks `big`